File: climatology/clim/orig/C/setupinterp.c
/* NAME gaussinterp (C version)
   16 Jan 00    earmstrong      NASA/JPL
   $Revision: 1.16 $

  DESCRIPTION
  Gaussian interpolation program modified to map SST (HDF) data.
  This program is a C port of the fortran version of gaussinterp.

  SYNOPSIS
  setupinterp() allocates space for arrays, calls interp()
  which does the interpolation

  USAGE
  % gaussinterp <infile> <beg_rec> <parmfile> <outfile> */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <libgen.h>
#include <math.h>
#include "interp.h"
/* see interp.h for global structs and arrays */
      
main( int argc, char *argv[] ) {

      int iday, iyr, calloc_bytes, rec_step;
      long max_infile_size;

      int value;
      FILE *fid_infile, *fid_parm;

      char datafile[RECSIZE];
      char *basefile;
      char *parmfile;
      char *startrec, *endrec;
      char cday[3], cyr[4], cdate[7];
      char *cstartdate;
      char *cenddate;
      char tmpline[80];

/* check command line arguments */
      if( argc != 5 && argc != 6 ) {
	  usage();
	  exit(1);
      }
      fileio.infile = argv[1];
      startrec = argv[2];
      parmfile = argv[3];
      fileio.outfile = argv[4];
      if( argc == 6 )
          fileio.aerosol_list = argv[5];  

      /* open parameter file and read line by line */
      if( (fid_parm = fopen( parmfile, "r" )) == NULL ){
	  printf( "\nCan not open %s\n", parmfile ); 
	  exit(1);
      }

      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%d", &parms.ndatamax ); 
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%d %d", &parms.x_length, &parms.y_length ); 
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%d", &parms.kmax ); 
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%f %f %f", &parms.xwin2, &parms.ywin2, &parms.zwin2 ); 
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%f %f %f", &parms.xh, &parms.yh, &parms.zh ); 
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%f %f %f", &parms.dx, &parms.dy, &parms.dz ); 
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%f %f", &parms.xlo, &parms.xhi ); 
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%f %f", &parms.ylo, &parms.yhi ); 
      fgets( fileio.mapformat, sizeof(fileio.mapformat), fid_parm );
      fgets( fileio.outformat, sizeof(fileio.outformat), fid_parm );
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%d", &cloud_erosion );
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%d", &sst_type );
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%d", &quality_flag );
      fgets( tmpline, sizeof(tmpline), fid_parm );
      sscanf( tmpline, "%d", &aerosol );

      fileio.mapformat[ strlen(fileio.mapformat)-1 ] = '\0';
      fileio.outformat[ strlen(fileio.outformat)-1 ] = '\0';

      fclose( fid_parm );
      
      if( strcmp(fileio.mapformat, "interleave") != 0  &&   
          strcmp(fileio.mapformat, "sequential") != 0 ) {
	   printf( "check parameter file for interleave or sequential maps !\n" );
	   exit(1);
      } 
      if( strcmp(fileio.outformat, "formatted") != 0  &&   
          strcmp(fileio.outformat, "unformatted") != 0 ) {
	   printf( "check parameter file for formatted or unformatted output !\n" );
	   exit(1);
      }

    /* Calculate number of interpolation "steps" based on
       long and lat ranges divided by interpolation interval.
       Origin is at lon=-180, lat=90 */
      parms.imax = floorf((parms.xhi + 180) / parms.dx) - floorf((parms.xlo + 180) / parms.dx );
      parms.jmax = floorf((parms.yhi - 90) / parms.dy) - floorf((parms.ylo - 90) / parms.dy );
      printf( "imax, jmax, kmax: %d %d %d \n", 
		parms.imax, parms.jmax, parms.kmax );

      parms.imin = 0;
      parms.jmin = 0;
      parms.kmin = 0;

    /* calculate degrees per grid point */
      parms.dxd =  360. / (float)parms.x_length;
      parms.dyd =  180. / (float)parms.y_length;

    /* find columns and rows for subsetting into hdf image array 
       remember @ hdfarray(0,0) -->  lon=-180, lat=90 */
      hdf.icolLeft  = floorf( (parms.xlo + 180.) / parms.dxd );
      hdf.icolRight = floorf( ((parms.xhi + 180.) / parms.dxd) - 1. );
      hdf.irowUpper = floorf( fabsf(parms.yhi - 90.) / parms.dyd );
      hdf.irowLower = floorf( (fabsf(parms.ylo - 90.) / parms.dyd) - 1. );

      /* printf( "array corners: %d %d %d %d \n", hdf.icolLeft, hdf.irowUpper, 
		  hdf.icolRight ,hdf.irowLower); */

   /* set global interpolation flag if necessary */
      if( parms.xlo == -180. && parms.xhi == 180. ) {
          all_zones = 1;  /* true */
          printf( "all_zones is true: interpolating across -180 West\n" );
      }

      if( cloud_erosion )
	  printf( "performing cloud erosion filtering on SST data \n" );
      if( sst_type )
	  printf( "working with 'all pixel' data: quality flag is %d \n", quality_flag );
      if( aerosol )
	  printf( "performing aerosol correction \n" );

   /* allocate space for arrays  */
      calloc_bytes = 0;
      if( (temp = fmal2d(  parms.x_length, parms.y_length )) == NULL ) 
	  printf( "not enough memory for array temp\n" ), exit(-1);
      else calloc_bytes = parms.x_length * parms.y_length * sizeof(float);  

      if( (in_data = cmal2d( parms.x_length, parms.y_length )) == NULL ) 
	  printf( "not enough memory for array in_data\n" ), exit(-1);
      else calloc_bytes += parms.x_length * parms.y_length * sizeof(char);  

      if( (quality_data = cmal2d( parms.x_length, parms.y_length )) == NULL ) 
	  printf( "not enough memory for array quality_data\n" ), exit(-1);
      else calloc_bytes += parms.x_length * parms.y_length * sizeof(char);  

       if( (sl = fmal3d( parms.imax, parms.jmax, parms.kmax )) == NULL )
	  printf( "not enough memory for array sl\n" ), exit(-1);
      else calloc_bytes += parms.imax * parms.jmax * parms.kmax * sizeof(float);  

      if( (vsum = fmal4d( 2, parms.imax, parms.jmax, parms.kmax )) == NULL )
	  printf( "not enough memory for array vsum\n" ), exit(-1);
      else calloc_bytes += parms.imax * parms.jmax * parms.kmax * 2 * sizeof(float);  
    
      if( (sst_seq = (float *)calloc( parms.imax, sizeof(float) )) == NULL ) 
	  printf( "not enough memory for array sst_seq\n" ), exit(-1);
      else calloc_bytes += parms.imax * sizeof(float);  

      if( (sst_inter = (float *)calloc( parms.kmax, sizeof(float) )) == NULL )
	  printf( "not enough memory for array sst_inter\n" ), exit(-1);
      else calloc_bytes += parms.kmax * sizeof(float);  

      if( (x = (float *)calloc( parms.ndatamax, sizeof(float) )) == NULL ) 
	  printf( "not enough memory for array x\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(float);  

      if( (y = (float *)calloc( parms.ndatamax, sizeof(float) )) == NULL )
	  printf( "not enough memory for array y\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(float);  

      if( (z = (float *)calloc( parms.ndatamax, sizeof(float) )) == NULL )
	  printf( "not enough memory for array z\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(float);  

      if( (f = (float *)calloc( parms.ndatamax, sizeof(float) )) == NULL )
	  printf( "not enough memory for array f\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(float);  

      /* Extend array xx by long_extend to handle crossing
	 longitude 180 during global interpolation.
         Extend by 4*search radius (plus 1 in xx dimension
         for -180 West crossing point */
      if( all_zones ) {
          long_extend = nint( 4 * parms.xwin2 / parms.dx );
          /* make sure long_extend is even for equal number of
	     bins on either side of -180 W  */
          if( long_extend % 2 != 0 )
              long_extend++;
          if( (xx = (float *)calloc( parms.imax + long_extend + 1, sizeof(float) )) == NULL )
	      printf( "not enough memory for array xx\n" ), exit(-1);
          else calloc_bytes += (parms.imax + long_extend + 1) * sizeof(float);
      } else {
             if( (xx = (float *)calloc( parms.imax, sizeof(float) )) == NULL )
	         printf( "not enough memory for array xx\n" ), exit(-1);
             else calloc_bytes += parms.imax * sizeof(float);
      }  
 
      if( (yy = (float *)calloc( parms.jmax, sizeof(float) )) == NULL)
	  printf( "not enough memory for array yy\n" ), exit(-1);
      else calloc_bytes += parms.jmax * sizeof(float);  

      if( (zz = (float *)calloc( parms.kmax, sizeof(float) )) == NULL )
	  printf( "not enough memory for array zz\n" ), exit(-1);
      else calloc_bytes += parms.kmax * sizeof(float);  

      if( (ixmin = (int *)calloc( parms.ndatamax, sizeof(int) )) == NULL ) 
	  printf( "not enough memory for array ixmin\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(int);  

      if( (ixmax = (int *)calloc( parms.ndatamax, sizeof(int) )) == NULL ) 
	  printf( "not enough memory for array ixmax\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(int);  

      if( (iymin = (int *)calloc( parms.ndatamax, sizeof(int) )) == NULL ) 
	  printf( "not enough memory for array iymin\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(int);  

      if( (iymax = (int *)calloc( parms.ndatamax, sizeof(int) )) == NULL ) 
	  printf( "not enough memory for array iymax\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(int);  

      if( (izmin = (int *)calloc( parms.ndatamax, sizeof(int) )) == NULL ) 
	  printf( "not enough memory for array izmin\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(int);  

      if( (izmax = (int *)calloc( parms.ndatamax, sizeof(int) )) == NULL ) 
	  printf( "not enough memory for array izmax\n" ), exit(-1);
      else calloc_bytes += parms.ndatamax * sizeof(int);  


      printf( "done calloc arrays\n" );  
      printf( "total bytes used: %d (%5.1f Mb)\n\n", calloc_bytes, calloc_bytes/1e6 ); 


      /* Determine zlo which corresponds to record number in infile for
	 first map center and number of days since 1 Jan 1985 */
      fileio.istart = atoi( startrec );
 
      /* Open input file list and read first record */
      if( (fid_infile = fopen( fileio.infile, "r" )) == NULL ){
	  printf( "\nCan not open %s\n", fileio.infile ); 
	  exit(1);
      }

      /* position to proper record in file */
      fseek( fid_infile, (fileio.istart-1)*RECSIZE, SEEK_SET );
      fgets( datafile, sizeof(datafile), fid_infile );
      fclose( fid_infile );
      basefile = basename( &datafile[0] );

      strcpy( cday, "   " );  /* initialize strings */
      strcpy( cyr, "    " ); 
      /* e.g, parse year and jul day from 1985004h54dd-gdm.hdf */
      strncpy( cyr, &basefile[0], 4 );
      iyr = atoi( cyr );
      strncpy( cday, &basefile[4], 3 );
      iday = atoi( cday );

      /* set parms.zlo: days since 1 Jan 1985 */
      parms.zlo = (float)iday+(float)(iyr-1985)*365.;
      if(iyr > 1988) parms.zlo = parms.zlo + 1.;
      if(iyr > 1992) parms.zlo = parms.zlo + 1.;
      if(iyr > 1996) parms.zlo = parms.zlo + 1.;
      if(iyr > 2000) parms.zlo = parms.zlo + 1.;
      if(iyr > 2004) parms.zlo = parms.zlo + 1.;
      if(iyr > 2008) parms.zlo = parms.zlo + 1.;
      printf( " zlo is  %f\n", parms.zlo );
      printf( " first map centered on day: %d in year: %d\n", iday, iyr );

      /* Set the first center dates */
      cstartdate = strcat( cdate, cyr );
      cstartdate = strcat( cdate, cday );
      date.startdate = atoi( cstartdate );
      printf( " start date is %d \n", date.startdate );

      /* reopen file to determine last map center date */
      fopen( fileio.infile, "r" );
      fileio.iend = fileio.istart + (parms.kmax-1) * parms.dz;
      fseek(fid_infile, (fileio.iend-1)*RECSIZE, SEEK_SET );
      fgets( datafile, sizeof(datafile), fid_infile );
      fclose( fid_infile );
      basefile = basename( &datafile[0] );
  
      /* parse year and jul day  */
      strcpy( cdate, "" );
      strncpy( cyr, &basefile[0], 4 );
      strncpy( cday, &basefile[4], 3 );
      cenddate = strcat( cdate, cyr );
      cenddate = strcat( cdate, cday );
      date.enddate = atoi( cenddate );
      printf( " end date is %d\n", date.enddate );

      /* Now determine temporal interpolation (record) ranges */
      /* set istart and iend so the first and last maps will span a complete
	 temporal interpolation range if possible. e.g., if first map is centered
	 on 1 Jan 1986 (this is date of first record to read from the 
	 input datafile [fileio.istart]), with a search radius of 5 days (zwin2),
         set istart back five days to 27 Dec 1985 and iend to 6 Jan 1986 */

      /* set initial step back and forward size to temporal search radius */ 
      rec_step = (int) parms.zwin2;  
            
      /* reopen input datafile */
      if( (fid_infile = fopen( fileio.infile, "r" )) == NULL ){
	  printf( "\nCan not open %s\n", fileio.infile ); 
	  exit(1);
      }

      /* determine max bytes in infile */  
      fseek( fid_infile, 0L, SEEK_END );
      max_infile_size = ftell( fid_infile );

      /* step back into infile; use return of fseek() to determine if 
         file pointer is set before beginning of file  */
      while( fseek(fid_infile, (fileio.istart-1-rec_step)*RECSIZE, SEEK_SET) != 0 ) 
          rec_step--;
      fileio.istart -= rec_step;

      /* step forward into infile */
      rec_step = parms.zwin2; 
      fseek( fid_infile, (fileio.iend-1+rec_step)*RECSIZE, SEEK_SET );

      /* if the file pointer is set beyond the EOF: step back */
      while( ftell(fid_infile) >= max_infile_size ) {
        rec_step--;  
	fseek( fid_infile, (fileio.iend-1+rec_step)*RECSIZE, SEEK_SET );
      }
      fileio.iend += rec_step;
 
      printf( " records: istart %d and iend %d \n\n", fileio.istart, fileio.iend );
      fclose( fid_infile );           
    
      /* call interp() */
      interp();
      
      free(temp), free(sl), free(vsum), free(sst_seq), free(sst_inter);
      free(x), free(y), free(z), free(f), free(xx), free(yy), free(zz);
      free(ixmin), free(ixmax), free(iymin), free(iymax);
      free(izmin), free(izmax), free(in_data), free(quality_data);
 }

/* functions to allocate space for float arrays */
/* 2d dimension */
float **fmal2d( int ix, int iy ){

 float **ival;
 int i;

 ival = (float **)calloc( ix, sizeof(float *) );
 for( i=0; i < ix; i++ )
     ival[i] = (float *)calloc( iy, sizeof(float) );

 return(ival);
}

 /* 3d dimension */
 float ***fmal3d( int ix, int iy, int iz ){

 float ***ival;
 int i, j;

 ival = (float ***)calloc( ix, sizeof(float **) );
 for( i=0; i < ix; i++ )
     ival[i] = (float **)calloc( iy, sizeof(float *) );

 for( i=0; i < ix; i++ )
     for( j=0; j < iy; j++ )
         ival[i][j] = (float *)calloc( iz, sizeof(float) );
 
 return(ival);
 }

/* 4d dimension */
float ****fmal4d( int ix, int iy, int iz, int ia ){
     
    float ****ival;
    int i, j, k;

    ival = (float ****)calloc( ix, sizeof(float ***) );
    for( i=0; i < ix; i++ )
	ival[i] = (float ***)calloc( iy, sizeof(float **) );
    for( i=0; i < ix; i++ )
	for( j=0; j < iy; j++)
	    ival[i][j] = (float **)calloc( iz, sizeof(float *) );   
    for( i=0; i < ix; i++ )
	for( j=0; j < iy; j++)
	    for( k=0; k < iz; k++)
	        ival[i][j][k] = (float *)calloc( ia, sizeof(float) );   

    return(ival);
}

/* function to allocate space for char array */
/* 2d dimension */
char **cmal2d( int ix, int iy ) {

 char **ival;
 int i;

 ival = (char **)calloc( ix, sizeof(char *) );
 for( i=0; i < ix; i++ )
     ival[i] = (char *)calloc( iy, sizeof(char) );

 return(ival);
}



usage() {
	  printf( "	~~ gaussian interpolation of pathfinder SST data (C version) ~~\n\n " );
	  printf(  "USAGE:  gaussinterp <infile> <beg_rec> <parmfile> <outfile>\n" );
	  printf(  " -- infile: file containing list to process\n" );
	  printf(  " -- beg_rec: first record (file) in infile processed\n" );
	  printf(  " -- parmfile: input parameter file (e.g., interp.parms)\n" );
	  printf(  " -- outfile: interpolated output file (ascii or binary)\n" );
	  printf(  "\n e.g.,  gaussinterp list.dat 100 interp.parms output.dat\n" );
	  printf(  "[process records of list.dat starting from record 100; interpolated \
result is output.dat]\n" );
	  printf(  "note: see interp.parms for example format of parameter file\n" );
	  printf(  "      see gaussinterp.readme for more general information\n" );
	  return(0);
}

File: climatology/clim/orig/Fortran/setupinterp.f
#include "interp.h"
c $Revision: 2.12 $
c NAME gaussinterp
c 7 Dec 99 	earmstrong 	NASA/JPL
c
c DESCRIPTION
c Gaussian interpolation program modified to map SST (HDF) data.
c Modified from intersstdbg7day3.f (jvazquez' prog).  
c
c SYNOPSIS
c setupinterp() allocates space for arrays, calls interp() which does the 
c interpolation
c  
c USAGE
c % gaussinterp infile beg_rec end_rec outfile parmfile

c **** Original program comments (intersstdbg7day3.f)  ***
c DESCRIPTION: Maps data when given the required information. Program
c   originally written by raghu. Version 7 reads the gridded data set.
c   Currently maps TMR water vapor.
c
c COMMENTS:
c  This programs wraps around in longitude for 0 to 360 maps.
c  If different maps size is needed in longitude w/o wraparound
c  some lines of code need to be modified in the file.
c
c
c CHANGES:
c  2/6/97 - change fint from integer*4 to integer*2; holds residuals
c 2/19/97 - change output to direct access file, change sl() to int*2
c 8/8/97 - change i/o so that it reads data directly from the residual
c          data files
c 9/9/97 - change main program so that i/o part are subroutines
c 9/10/97 - add header,time,lat,lon and #of maps to output file
c 9/11/97 - add version and output file specification in runtime
c           argument list
c **** end Orginal comments ****

c 12/01/99 - major modifications (ema). 
c	   	- added command line read for input, output filenames
c		  and records numbers of input file read
c		- binsum is now a subprogram
c		- mapping params (interp.h) read in with include statement
c		- removed some superfluous do loops
c 12/06/99 - dynamic allocation of arrays added
c	   - read input parms from arg 5 (see 'interp.parms'
c	     for an example of format)
c	   - hdf data array size still hardcoded in interp.h
c 12/10/99 - variable zlo set equal to command line arg for first
c	     record number in infile to process (istart). zlo 
c	     removed from parameter file (interp.parms)
c 12/13/99 - imax, jmax calculated from xlo, xhi, ylo, and yhi
c	     imin, jmin, kmin set to 1
c 12/16/99 - added options in parameter file for sequential
c	     or interleaved maps, and formatted or unformatted
c	     output file
c 12/27/99 - added basename C call. program can now handle 
c            compressed datafiles.
c 1/4/00   - added subsetting of hdf array based on lon/lat 
c	     bounds. zlo now read from parameter file. 
c 1/4/00   - zlo found from parse of first datafile.
c	     geo-positions of UL, LR written to output header

      program setupinterp
      
      integer ndatamax, x_length, y_length
      integer temp_malloc, sst_inter_malloc
      integer sst_seq_malloc, sl_malloc
      integer x_malloc, y_malloc, z_malloc, f_malloc
      integer xx_malloc, yy_malloc, zz_malloc, vsum_malloc
      integer ixmin_malloc, ixmax_malloc
      integer iymin_malloc, iymax_malloc
      integer izmin_malloc, izmax_malloc
      real*4 float_size
      integer*4 int_size

      character*80 plabel
      character*50 infile
      character*30 outfile
      character*60 parmfile
      character*10 mapformat
      character*15 outformat
      character*6 startrec
      character*6 endrec
      character*3 cday
      character*4 cyr
      character*150 datafile
      character*30 basename

      common /parms/imin,jmin,kmin,
     &	             xwin2,ywin2,zwin2,
     &		     xh,yh,zh,		
     &		     dx,dy,dz,
     &		     xlo,xhi,ylo,yhi,zlo,
     &		     dxd,dyd

      common /fileio/ infile,outfile,istart,
     &	              iend,mapformat,outformat	
      common /hdfindex/ icolLeft,icolRight,irowUpper,irowLower
       
      common /basefile/ basename

c ...check command line arguments
      call getarg( 1, infile )
      call getarg( 2, startrec )
      call getarg( 3, endrec )
      call getarg( 4, parmfile )
      call getarg( 5, outfile )

      inum_args = iargc()
      if ( inum_args .ne. 5 ) then
	  call usage()
	  stop
      endif

      read( startrec, '(i)' ) istart
      read( endrec, '(i)' ) iend

      if ( iend .lt. istart ) then
	  write(6,*) '\n	Error: end record number
     & smaller than beginning record ! \n'
	  stop
      endif

c--  zlo determined by reading first record and parsing the datafile
c--  for year and julian day using getbase()  (C call)

c-- Open input file list
      open( unit=ifile,status='old',file=infile,access='direct',
     &  recl=RECSIZE,form='formatted',err=2000,iostat=ierr )

       read( ifile, '(a)', rec=istart ) datafile
       close( ifile, status='keep', err=2500, iostat=ierr )

       call getbase( datafile )	! returns basename
       cday=basename( 5:7 )
       cyr=basename( 1:4 )
       read( cday,'(i3)' ) iday
       read( cyr,'(i4)' ) iyr
       zlo = float(iday)+float(iyr-1985)*365.
       if(iyr.gt.1988) zlo = zlo + 1.
       if(iyr.gt.1992) zlo = zlo + 1.
       if(iyr.gt.1996) zlo = zlo + 1.
       if(iyr.gt.2000) zlo = zlo + 1.
       if(iyr.gt.2004) zlo = zlo + 1.
       print *, 'zlo is ', zlo

c-- read in the parameter file 
      open( unit=iparm,status='old',form='formatted',file=parmfile,
     &     err=1000,iostat=ierr )

      read( unit=iparm, fmt=* ) ndatamax
      read( unit=iparm, fmt=* ) x_length, y_length
      read( unit=iparm, fmt=* ) kmax
      read( unit=iparm, fmt=* ) xwin2, ywin2, zwin2
      read( unit=iparm, fmt=* ) xh, yh, zh
      read( unit=iparm, fmt=* ) dx, dy, dz
      read( unit=iparm, fmt=* ) xlo, xhi
      read( unit=iparm, fmt=* ) ylo, yhi
c      read( unit=iparm, fmt=* ) zlo
c      read( unit=iparm, fmt=* ) dxd, dyd
      read( unit=iparm, fmt='(a)' ) mapformat
      read( unit=iparm, fmt='(a)' ) outformat
      close( iparm,status='keep',err=1500,iostat=ierr )

c	 write(*,*) ndatamax
c	 write(*,*) x_length, y_length
c	 write(*,*) kmax
c	 write(*,*) xwin2,ywin2, zwin2
c	 write(*,*) xh, yh, zh
c	 write(*,*) dx, dy, dz
c	 write(*,*) xlo, xhi
c	 write(*,*) ylo, yhi
c	 write(*,*) zlo
cc	 write(*,*) dxd, dyd
c	 write(*,*) mapformat
c	 write(*,*) outformat

      if( (mapformat .ne. 'interleave') .and.  
     &    (mapformat .ne. 'sequential') ) then
	   print *, 'check parameter file for interleave or sequential maps !'
	   stop
      endif
      if( (outformat .ne. 'formatted') .and.  
     &    (outformat .ne. 'unformatted') ) then
	   print *, 'check parameter file for formatted or unformatted output !'
	   stop
      endif


c-- calculate max number of interpolation "steps" based on
c-- long and lat ranges divided by interpolation interval
c      imax = ( (xhi - xlo) + 1 ) / dx
c      jmax = ( abs(yhi - ylo) + 1 ) / dy
      imax = ( abs(xhi - xlo)  ) / dx
      jmax = ( abs(yhi - ylo)  ) / dy

      print *, 'imax, jmax, kmax',  imax,jmax,kmax

      imin = 1
      jmin = 1
      kmin = 1

c-- calculate degrees per grid point
      dxd =  360. / float(x_length)
      dyd =  180. / float(y_length)

c-- find columns and rows for subsetting into hdf image array 
c-- remember @ hdfarray(1,1) -->  lon=-180, lat=90
      icolLeft = nint( ((xlo + 180) / dxd) + 1 )
      icolRight = nint( (xhi + 180) / dxd )
      irowUpper = nint( (abs(yhi - 90) / dyd) + 1 )
      irowLower = nint( (abs(ylo - 90) / dyd) )

c      print *, 'icolLeft, icolRight,irowUpper, irowLower', 
c     & icolLeft, icolRight, irowUpper, irowLower 

c-- 4 bytes for floats and ints
      float_size = 4
      int_size = 4

c-- allocate space for arrays
          itotsize = x_length*y_length*float_size
      temp_malloc = malloc( %val(itotsize) )     
          itotsize = kmax*float_size
      sst_inter_malloc = malloc( %val(itotsize) )     
          itotsize = imax*float_size
      sst_seq_malloc = malloc( %val(itotsize) )     
          itotsize = imax*jmax*kmax*float_size
      sl_malloc = malloc( %val(itotsize) ) 
          itotsize =  ndatamax*float_size
      x_malloc = malloc( %val(itotsize) ) 
      y_malloc = malloc( %val(itotsize) ) 
      z_malloc = malloc( %val(itotsize) ) 
      f_malloc = malloc( %val(itotsize) ) 
          itotsize = imax*float_size
      xx_malloc = malloc( %val(itotsize) ) 
          itotsize = jmax*float_size
      yy_malloc = malloc( %val(itotsize) ) 
          itotsize = kmax*float_size
      zz_malloc = malloc( %val(itotsize) ) 
          itotsize = 2*imax*jmax*kmax*float_size
      vsum_malloc = malloc( %val(itotsize) ) 
          itotsize = ndatamax*int_size
      ixmin_malloc = malloc( %val(itotsize) ) 
      ixmax_malloc = malloc( %val(itotsize) ) 
      iymin_malloc = malloc( %val(itotsize) ) 
      iymax_malloc = malloc( %val(itotsize) ) 
      izmin_malloc = malloc( %val(itotsize) ) 
      izmax_malloc = malloc( %val(itotsize) ) 

c      print *, 'done malloc arrays\n'

c--- call interp() with 'space' variables passed call by value
      call interp( %val(temp_malloc), 
     &      %val(sst_inter_malloc), %val(sst_seq_malloc),
     &      %val(sl_malloc), %val(x_malloc), %val(y_malloc), 
     &	    %val(z_malloc), %val(f_malloc), %val(xx_malloc),
     &	    %val(yy_malloc), %val(zz_malloc), 
     &	    %val(vsum_malloc), %val(ixmin_malloc), %val(ixmax_malloc),
     &	    %val(iymin_malloc), %val(iymax_malloc), %val(izmin_malloc), 
     &	    %val(izmax_malloc), ndatamax, x_length, y_length,
     &	    imax, jmax, kmax )

c-- how to free memory? dunno


 1000 print *, 'Error opening parameter file: ', parmfile, 'error num: ', ierr
 1500 print *, 'Error closing parameter file: ', parmfile, 'error num: ', ierr
 2000 print *, 'Error opening input: ', infile, 'error num: ', ierr
 2500 print *, 'Error closing input: ', infile, 'error num: ', ierr

      end

      subroutine usage()
	  print *, '\n	~~ gaussian interpolation of pathfinder SST data
     & (f77 version) ~~\n '
	  print *, 'USAGE: % gaussinterp infile beg_rec end_rec parmfile outfile\n'
	  print *, ' -- infile: file containing list to process
     & (ascii; fixed length records)'
	  print *, ' -- beg_rec: first record (file) in infile processed'
	  print *, ' -- end_rec: last record (file) in infile processed'
	  print *, ' -- parmfile: input parameter file (e.g., interp.parms)'
	  print *, ' -- outfile: interpolated output file (ascii or binary)'
	  print *, '\n e.g., % gaussinterp list.dat 100 600 interp.parms output.dat '
	  print *, '[process records 100 to 600 of list.dat; interpolated
     & result is output.dat]\n'
	  print *, 'note: see interp.parms for example format of parameter file'
	  print *, '      see gaussinterp.readme for more general information'
	  print *, '      this executable compiled for image x and y size:', XSIZE, YSIZE  
      end
